### ISSUE TITLE
[BUG] Hot Reload is unable to find changes made to WeatherForecastController.cs

### ISSUE THREAD

**Issue by RobertOstermann** [microsoft/vscode-dotnettools#794]
### Describe the Issue

I am using .NET 8 on Ubuntu 22.04 on WSL. The hot reload symbol shows in the debug bar, but clicking on the hot reload button does nothing. The `C# Hot Reload` output window shows the following.

```
💡 Possible changes were found.
Checking for updates...
No changes were found.
```

The debugger seems to not recognize changes in the file, even though I am making changes. I have not tested on Windows or Mac.

Additionally, in the debug console I see the following message

```
Breakpoint warning: The source code is different from the original version.
To allow this breakpoint to be hit: Add '"requireExactSource": false' to launch.json and restart debugging.
```

Changing this setting removes the message, but does not allow the hot reload to work.

### Steps To Reproduce

1. On a Windows machine use WSL (Ubuntu 22.04)
2. Setup a C# web api
3. Setup a basic launch.json file for C# debugging
4. Run the project and make changes to the code
5. Save the file -the hot reload does not work
6. Click the hot reload button - the hot reload still does not work.

### Expected Behavior

The changes should force a hot reload.

### Environment Information

- OS - Linux (WSL)
- VSCode - 1.86.0-insider
- C# Dev Kit - v1.2.2

**Comment by RobertOstermann** [comment]
Any update on this?

If it helps I tried this on Windows and experienced the same bug.
Here is my launch.json setup
```json
{
  "name": "Server (api) - Windows",
  "type": "dotnet",
  "request": "launch",
  "projectPath": "${workspaceFolder}\\server\\server.csproj",
  "launchConfigurationId": "TargetFramework=;api",
}
```

**Comment by isadorasophia** [comment]
Thanks for the ping! We determined that the root cause issue was how C# Dev Kit was initializing some of its services. The fix should be available in the latest **v1.3.2** extension! Please let me know if that works for you.

**Comment by RobertOstermann** [comment]
@isadorasophia I am still experiencing this issue after updating to v1.3.2 of the extension. There is a new part of the message which mentions hot reload is unavailable for the session. I am seeing this for an ASP.NET Core Web API, I tried with both minimal api and with controllers.

```
Starting session.
❗ Hot Reload is unavailable for this session.
💡 Possible changes were found.
Checking for updates...
No changes were found.
```

**Comment by isadorasophia** [comment]
Thanks Robert, I can confirm and repro that the message `❗ Hot Reload is unavailable for this session.` is indeed an error on our side, as hot reload is initialized afterwards.

I will try the ASP.NET Core Web API. Is this when editing a `.cs` file?

**Comment by RobertOstermann** [comment]
Yes. This is when editing a .cs file.

**Comment by RobertOstermann** [comment]
@isadorasophia Any update on this?

**Comment by Shepherdsinan** [comment]
![image](https://github.com/microsoft/vscode-dotnettools/assets/65864888/25c9307a-2f2b-4028-9b15-a38406717ac6)

**Comment by RobertOstermann** [comment]
@isadorasophia Should I create a new issue to get this looked at?

**Comment by isadorasophia** [comment]
Hi @RobertOstermann, we are still taking a look at this. We can keep using this issue to track it down. The "Hot Reload is unavailable for this session" error message should have been addressed as of C# Dev Kit `v1.3.8` and C# extension `v2.16.24`. 

However, I am no longer able to reproduce the "No changes were found" on my side, so I have a couple of follow-up questions:
- Does this only reproduce when targeting WSL? Does it also reproduce on Windows (or Linux/macOS)?
- Does this also reproduce on a console app? Or is it only when targeting ASP.NET applications?

**Comment by RobertOstermann** [comment]
@isadorasophia This reproduces on Windows 10, I have only tested on Windows 10 (WSL and regular). This reproduces both on the console app and on web api, though I am not seeing "Hot Reload is unavailable for this session" anymore.

Console App: Says that updates were applied, but the program does not change and the hot reload does not work. In this example I started the program with the `Console.WriteLine("write line")` not commented out, then I input "test", the ouput in the debug console was as expected, then I commented out that line and input "test" again, the output does not show correctly and it I was no longer able to hit a breakpoint.

![image](https://github.com/microsoft/vscode-dotnettools/assets/46828475/56566285-178d-47a4-ac9d-c9bd2a0ca016)

Web API: Does not find the changes. I created a basic web api in Visual Studio, then switched to VSCode and added the line `Console.WriteLine("test")`, I hit the endpoint and it worked as expected, then I commented out that line and got the "Breakpoint warning" showing in the debug console.

![image](https://github.com/microsoft/vscode-dotnettools/assets/46828475/a5f09f09-989e-4952-9577-623602e2c7bc)

As @Shepherdsinan seems to be having the same issue, perhaps he could provide additional information.

**Comment by Shepherdsinan** [comment]
My project .net core version 8 not working hot reload

**Comment by Shepherdsinan** [comment]
> Hi @RobertOstermann, we are still taking a look at this. We can keep using this issue to track it down. The "Hot Reload is unavailable for this session" error message should have been addressed as of C# Dev Kit `v1.3.8` and C# extension `v2.16.24`. 
> 
> However, I am no longer able to reproduce the "No changes were found" on my side, so I have a couple of follow-up questions:
> - Does this only reproduce when targeting WSL? Does it also reproduce on Windows (or Linux/macOS)?
> - Does this also reproduce on a console app? Or is it only when targeting ASP.NET applications?

C# Dev kit v1.3.8 and C# extension hot reload work thanks 🙏

**Comment by isadorasophia** [comment]
@RobertOstermann Very helpful, thanks for the screenshots!

Regarding the Console App scenario: when applying changes with hot reload, if your application was running (not stopped at a breakpoint), the change will only be executed once that frame reenters the stack. I also might have misunderstood from your screenshot, so I'll confirm just in case. 

In order to hit the breakpoint to the code you just edited and to reenter the code, you either need a code like this:

```csharp
Console.WriteLine("hi");

string? line;
while ((line = Console.ReadLine()) != null)
{
    Foo();
}

static void Foo()
{
    // changes here will be immediately executed because Foo() reenters the stack every loop
    Console.WriteLine("read line"); 
    Console.WriteLine("write line");
}
```

or this:
```csharp
Console.WriteLine("hi");

string? line;
while ((line = Console.ReadLine()) != null)
{
    // only make changes after the debugger stops at this line below.
    // it will make sure that the instruction is remapped to be executed at the new code you just wrote.
    System.Diagnostics.Debugger.Break(); 

    Console.WriteLine("read line!");
    Console.WriteLine("write line");
}
```

Regarding the WebApi scenario: I was able to reproduce the problem! Apparently edits to `WeatherForecastController.cs` _specifically_ are not being picked up. I am not sure why that is the case yet. I'm following up with @tmat, since this is likely on how Roslyn is picking up changes.

**Comment by RobertOstermann** [comment]
@isadorasophia The console application is working as expected according to your explanation.

**Comment by isadorasophia** [comment]
I have reopened the issue so we can track the WebApi scenario

**Comment by sig00x** [comment]
Can confirm the same issue in WebAPI with c# dev kit v1.3.10 and c# v2.15.30.
Hot reload does absolutely nothing. As soon as I change the code while running the breakpoints don't get hit anymore.

**Comment by givemeurhats** [comment]
> Regarding the WebApi scenario: I was able to reproduce the problem! Apparently edits to `WeatherForecastController.cs` _specifically_ are not being picked up.
I made a new web API net 8 project and replaced the default WeatherForecastController with a custom one but the issue still exists. In my tests it's not _specifically_ `WeatherForecastController.cs`.

**Comment by RobertOstermann** [comment]
@isadorasophia Is there any update on this issue?

**Comment by tr1age** [comment]
Still an issue here as well.

**Comment by RobertOstermann** [comment]
@isadorasophia Any update on this? This has been an issue for several months and makes debugging with hot reload impossible. Seem it might also be related to #1034

**Comment by ryzngard** [comment]
This should be working. Verified with c# `v2.32.14 (pre-release)`. Let me know if you hit any issues!

**Comment by RobertOstermann** [comment]
@ryzngard This is still not working. @tr1age mentioned it was an issue 1 hour ago and I just tested on the latest pre-release versions of the extensions. Hot reload does not happen, as described in earlier comments.

**Comment by ryzngard** [comment]
Do you have a repro project I can test with? It was working for me the few apps I tried. Alternatively, you can file a bug through vs code with the `Razor: Report a Razor Issue` command as well.

**Comment by ryzngard** [comment]
I should clarify one thing: `.razor` and `.cshtml` file hot reload is still blocked on https://github.com/dotnet/razor/issues/10341 , but `.cs` files for blazor server apps should work

**Comment by ryzngard** [comment]
Ah, I found a repro in a project. Reopening while I investigate, feel free to add more information as it comes

**Comment by RobertOstermann** [comment]
@ryzngard It appears using `"type": "coreclr"` instead of `"type": "dotnet"` is a workaround. Not sure why coreclr would work but not dotnet, but that is what I have seen when experimenting with it.

**Comment by mariusdumitru** [comment]
If you encounter an issue in Visual Studio where Hot Reload is not working in your Blazor project, and you see the message "No code changes were found" even after making changes, here is a possible solution:

Problem:
A line in your .csproj file might be causing the Hot Reload feature to fail. Specifically, the following line could have been added unintentionally:
<UseRazorSourceGenerator>false</UseRazorSourceGenerator>

Solution:
Open your Blazor project’s .csproj file.
Look for the line:
<UseRazorSourceGenerator>false</UseRazorSourceGenerator>

Remove this line completely from the .csproj file. Setting it to true does not resolve the issue, but removing it should fix the problem.
Save the .csproj file and reload the project.
After removing the line, Hot Reload should work as expected without displaying the "No code changes were found" message.

**Comment by webreidi** [comment]
@ryzngard Is this still reproducible?

**Comment by ryzngard** [comment]
@webreidi yes, it's on my current list of items now that FUSE work is wrapping up

**Comment by Eptagone** [comment]
This is still happening :(
I just created an empty web project and edit the msg from `Program.cs`

**Comment by sid198315** [comment]
Hello @ryzngard - Is this still on the radar? We have shifted from VS Enterprise to VS Code. We can still easily reproduce this issue. Do you have an estimate when this will be fixed. C# Dev Kit version - 1.80.7
